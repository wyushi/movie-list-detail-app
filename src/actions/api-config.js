export const REQUEST_API_CONFIG = 'REQUEST_API_CONFIG'
export const RECEIVE_API_CONFIG = 'RECEIVE_API_CONFIG'

const requestAPIConfig = () => {
  return {
    type: REQUEST_API_CONFIG
  }
}

const receiveAPIConfig = (json) => {
  return {
    type: RECEIVE_API_CONFIG,
    apiConfig: json,
    receivedAt: Date.now()
  }
}

export const fetchAPIConfig = () => {
  return dispatch => {
    dispatch(requestAPIConfig())
    return fetch(url)
      .then(response => response.json())
      .then(json => dispatch(receiveAPIConfig(json)))
  }
}