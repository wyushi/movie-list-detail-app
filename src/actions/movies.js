export const REQUEST_MOVIES = 'REQUEST_MOVIES'
export const RECEIVE_MOVIES = 'RECEIVE_MOVIES'

const url = 'https://api.themoviedb.org/3/discover/movie?sort_by=popularity.desc&api_key=3403673c3e3cf351c8741bda8180865e'

const requestMovies = (page) => {
  return {
    type: REQUEST_MOVIES,
    page
  }
}

const receiveMovies = (json) => {
  return {
    type: RECEIVE_MOVIES,
    movies: json,
    receivedAt: Date.now()
  }
}

export const fetchMovies = () => {
  return dispatch => {
    dispatch(requestMovies())
    return fetch(url)
      .then(response => response.json())
      .then(json => dispatch(receiveMovies(json)))
  }
}