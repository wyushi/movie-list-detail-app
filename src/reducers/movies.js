import {
  REQUEST_MOVIES,
  RECEIVE_MOVIES
} from 'actions/movies'

export const initialState = {
  isFetching: false,
  page: 1,
  results: [],
  error: null
}

const movies = (state=initialState, action) => {
  switch(action.type) {
    case REQUEST_MOVIES:
      return handleRequestMoviesAction(state, action)
    case RECEIVE_MOVIES:
      return handleReceiveMoviesAction(state, action)
    default:
      return state
  }
}

const handleRequestMoviesAction = (state, action) => {
  return {
    ...state,
    isFetching: true,
    page: action.page, 
    error: null
  }
}

const handleReceiveMoviesAction = (state, action) => {
  if (!action.movies.results) {
    return {
      ...state,
      isFetching: false,
      error: {
        ...action.movies,
        message: JSON.stringify(action.movies)
      }
    }
  }

  const { page } = action.movies
  const results = page === 1 ?
    action.movies.results : [
      ...state.results,
      ...action.movies.results ]
  return {
    ...state,
    isFetching: false,
    page,
    results,
    error: null
  }
}

export default movies