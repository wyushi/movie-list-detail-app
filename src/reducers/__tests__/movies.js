import moives, { initialState } from '../movies'
import { requestMovies, receiveMovies } from '../../actions/movies'
import data from '../../config/jest/mockData'

it('returns the same state on an unhandled action', () => {
  expect(moives(initialState, {type: '_NULL'})).toEqual(initialState)
})

it('handle a REQUEST_MOVIES action', () => {
  expect(moives(initialState, requestMovies(2))).toEqual({
    ...initialState,
    isFetching: true,
    page: 2
  })
})

it('handle a RECEIVE_MOVIES action', () => {
  const state = moives(initialState, receiveMovies(data))
  expect(state).toMatchSnapshot()
  expect(state.isFetching).toBe(false)
  expect(state.results).toBeTruthy()
  expect(state.results.length).toEqual(20)
})