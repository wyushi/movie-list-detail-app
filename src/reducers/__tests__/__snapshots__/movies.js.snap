// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handle a RECEIVE_MOVIES action 1`] = `
Object {
  "error": null,
  "isFetching": false,
  "page": 1,
  "results": Array [
    Object {
      "adult": false,
      "backdrop_path": "/5A2bMlLfJrAfX9bqAibOL2gCruF.jpg",
      "genre_ids": Array [
        28,
        14,
        878,
        12,
      ],
      "id": 297802,
      "original_language": "en",
      "original_title": "Aquaman",
      "overview": "Once home to the most advanced civilization on Earth, the city of Atlantis is now an underwater kingdom ruled by the power-hungry King Orm. With a vast army at his disposal, Orm plans to conquer the remaining oceanic people -- and then the surface world. Standing in his way is Aquaman, Orm's half-human, half-Atlantean brother and true heir to the throne. With help from royal counselor Vulko, Aquaman must retrieve the legendary Trident of Atlan and embrace his destiny as protector of the deep.",
      "popularity": 805.109,
      "poster_path": "/5Kg76ldv7VxeX9YlcQXiowHgdX6.jpg",
      "release_date": "2018-12-07",
      "title": "Aquaman",
      "video": false,
      "vote_average": 6.9,
      "vote_count": 3078,
    },
    Object {
      "adult": false,
      "backdrop_path": "/hMANgfPHR1tRObNp2oPiOi9mMlz.jpg",
      "genre_ids": Array [
        28,
        12,
        878,
      ],
      "id": 424783,
      "original_language": "en",
      "original_title": "Bumblebee",
      "overview": "On the run in the year 1987, Bumblebee finds refuge in a junkyard in a small Californian beach town. Charlie, on the cusp of turning 18 and trying to find her place in the world, discovers Bumblebee, battle-scarred and broken.  When Charlie revives him, she quickly learns this is no ordinary yellow VW bug.",
      "popularity": 296.914,
      "poster_path": "/fw02ONlDhrYjTSZV8XO6hhU3ds3.jpg",
      "release_date": "2018-12-15",
      "title": "Bumblebee",
      "video": false,
      "vote_average": 6.5,
      "vote_count": 999,
    },
    Object {
      "adult": false,
      "backdrop_path": "/93xA62uLd5CwMOAs37eQ7vPc1iV.jpg",
      "genre_ids": Array [
        18,
        10402,
      ],
      "id": 424694,
      "original_language": "en",
      "original_title": "Bohemian Rhapsody",
      "overview": "Singer Freddie Mercury, guitarist Brian May, drummer Roger Taylor and bass guitarist John Deacon take the music world by storm when they form the rock 'n' roll band Queen in 1970. Hit songs become instant classics. When Mercury's increasingly wild lifestyle starts to spiral out of control, Queen soon faces its greatest challenge yet – finding a way to keep the band together amid the success and excess.",
      "popularity": 257.009,
      "poster_path": "/gbmkFWdtihe1VfydTDsieQ6VfGL.jpg",
      "release_date": "2018-10-24",
      "title": "Bohemian Rhapsody",
      "video": false,
      "vote_average": 8.2,
      "vote_count": 4208,
    },
    Object {
      "adult": false,
      "backdrop_path": "/6OTRuxpwUUGbmCX3MKP25dOmo59.jpg",
      "genre_ids": Array [
        28,
        16,
        878,
      ],
      "id": 503314,
      "original_language": "ja",
      "original_title": "ドラゴンボール超スーパー ブロリー",
      "overview": "Earth is peaceful following the Tournament of Power. Realizing that the universes still hold many more strong people yet to see, Goku spends all his days training to reach even greater heights. Then one day, Goku and Vegeta are faced by a Saiyan called 'Broly' who they've never seen before. The Saiyans were supposed to have been almost completely wiped out in the destruction of Planet Vegeta, so what's this one doing on Earth? This encounter between the three Saiyans who have followed completely different destinies turns into a stupendous battle, with even Frieza (back from Hell) getting caught up in the mix.",
      "popularity": 209.298,
      "poster_path": "/f03YksE4NggUjG75toz4H1YAGRf.jpg",
      "release_date": "2018-12-14",
      "title": "Dragon Ball Super: Broly",
      "video": false,
      "vote_average": 7.5,
      "vote_count": 276,
    },
    Object {
      "adult": false,
      "backdrop_path": "/8yqLPNwNCtpOPc3XkOlkSMnghzw.jpg",
      "genre_ids": Array [
        18,
        28,
      ],
      "id": 480530,
      "original_language": "en",
      "original_title": "Creed II",
      "overview": "Follows Adonis Creed's life inside and outside of the ring as he deals with new found fame, issues with his family, and his continuing quest to become a champion.",
      "popularity": 205.19,
      "poster_path": "/v3QyboWRoA4O9RbcsqH8tJMe8EB.jpg",
      "release_date": "2018-11-21",
      "title": "Creed II",
      "video": false,
      "vote_average": 6.6,
      "vote_count": 807,
    },
    Object {
      "adult": false,
      "backdrop_path": "/9cmrRktcAmhteKMNRl2hU6rQzLJ.jpg",
      "genre_ids": Array [
        27,
        53,
        28,
      ],
      "id": 522681,
      "original_language": "en",
      "original_title": "Escape Room",
      "overview": "Six strangers find themselves in circumstances beyond their control, and must use their wits to survive.",
      "popularity": 199.039,
      "poster_path": "/8yZAx7tlKRZIg7pJfaPhl00yHIQ.jpg",
      "release_date": "2019-01-03",
      "title": "Escape Room",
      "video": false,
      "vote_average": 6.5,
      "vote_count": 67,
    },
    Object {
      "adult": false,
      "backdrop_path": "/rxYG6Sj95as9rv9wKIHUx6ATWd3.jpg",
      "genre_ids": Array [
        878,
        28,
        12,
        14,
        53,
      ],
      "id": 428078,
      "original_language": "en",
      "original_title": "Mortal Engines",
      "overview": "Set in a world many thousands of years in the future. Earth’s cities now roam the globe on huge wheels, devouring each other in a struggle for ever diminishing resources. On one of these massive Traction Cities, Tom Natsworthy has an unexpected encounter with a mysterious young woman from the Outlands who will change the course of his life forever.",
      "popularity": 190.13,
      "poster_path": "/gLhYg9NIvIPKVRTtvzCWnp1qJWG.jpg",
      "release_date": "2018-12-05",
      "title": "Mortal Engines",
      "video": false,
      "vote_average": 6,
      "vote_count": 632,
    },
    Object {
      "adult": false,
      "backdrop_path": "/71OjxI27tK7kTIiPnMkdQDx14pe.jpg",
      "genre_ids": Array [
        80,
        18,
        53,
      ],
      "id": 401469,
      "original_language": "en",
      "original_title": "Widows",
      "overview": "A police shootout leaves four thieves dead during an explosive armed robbery attempt in Chicago. Their widows have nothing in common except a debt left behind by their spouses' criminal activities. Hoping to forge a future on their own terms, they join forces to pull off a heist.",
      "popularity": 190.008,
      "poster_path": "/tvmPiTShsfp4vSUBFsCHYaX9M7i.jpg",
      "release_date": "2018-11-06",
      "title": "Widows",
      "video": false,
      "vote_average": 6.8,
      "vote_count": 406,
    },
    Object {
      "adult": false,
      "backdrop_path": "/u8CP7EeWbYMlIVqEoKAt6OYsEe1.jpg",
      "genre_ids": Array [
        28,
        80,
        35,
        18,
      ],
      "id": 483906,
      "original_language": "en",
      "original_title": "Polar",
      "overview": "When a retiring assassin realizes he is the target of a hit, he winds up back in the game going head to head with a gang of younger, ruthless killers.",
      "popularity": 172.028,
      "poster_path": "/qOBEpKVLl8Q9CZScbOcRRVISezV.jpg",
      "release_date": "2019-01-25",
      "title": "Polar",
      "video": false,
      "vote_average": 6.5,
      "vote_count": 172,
    },
    Object {
      "adult": false,
      "backdrop_path": "/wqtaHWOEZ3rXDJ8c6ZZShulbo18.jpg",
      "genre_ids": Array [
        18,
        10402,
        10749,
      ],
      "id": 332562,
      "original_language": "en",
      "original_title": "A Star Is Born",
      "overview": "Seasoned musician Jackson Maine discovers—and falls in love with—struggling artist Ally. She has just about given up on her dream to make it big as a singer—until Jack coaxes her into the spotlight. But even as Ally's career takes off, the personal side of their relationship is breaking down, as Jack fights an ongoing battle with his own internal demons.",
      "popularity": 169.781,
      "poster_path": "/wrFpXMNBRj2PBiN4Z5kix51XaIZ.jpg",
      "release_date": "2018-10-03",
      "title": "A Star Is Born",
      "video": false,
      "vote_average": 7.5,
      "vote_count": 3040,
    },
    Object {
      "adult": false,
      "backdrop_path": "/bkc4AY6FTa3yNqrshE9b1elDzPt.jpg",
      "genre_ids": Array [
        80,
        18,
        53,
      ],
      "id": 504172,
      "original_language": "en",
      "original_title": "The Mule",
      "overview": "A 90 year old horticulturalist and Korean War veteran is caught transporting $3 million worth of cocaine through Illinois for a Mexican drug cartel.",
      "popularity": 168.38,
      "poster_path": "/t0idiLMalKMj2pLsvqHrOM4LPdQ.jpg",
      "release_date": "2018-12-14",
      "title": "The Mule",
      "video": false,
      "vote_average": 6.3,
      "vote_count": 270,
    },
    Object {
      "adult": false,
      "backdrop_path": "/AuA50D7T7S7OEVcGo0ZKaMhERn0.jpg",
      "genre_ids": Array [
        12,
        28,
        53,
      ],
      "id": 375588,
      "original_language": "en",
      "original_title": "Robin Hood",
      "overview": "A war-hardened Crusader and his Moorish commander mount an audacious revolt against the corrupt English crown.",
      "popularity": 144.527,
      "poster_path": "/AiRfixFcfTkNbn2A73qVJPlpkUo.jpg",
      "release_date": "2018-11-20",
      "title": "Robin Hood",
      "video": false,
      "vote_average": 5.8,
      "vote_count": 462,
    },
    Object {
      "adult": false,
      "backdrop_path": "/3haESpzscyP1pLl2BxQ9j5KK9cM.jpg",
      "genre_ids": Array [
        16,
        28,
        14,
        12,
      ],
      "id": 390634,
      "original_language": "ja",
      "original_title": "劇場版「Fate/stay night [Heaven's Feel] ⅠⅠ. lost butterfly」",
      "overview": "Theatrical-release adaptation of the visual novel \\"Fate/stay night\\", following the third and final route. (Part 2 of a trilogy.)",
      "popularity": 138.227,
      "poster_path": "/t0jwlwFUrtkPJoq5JH9YrRVROXZ.jpg",
      "release_date": "2019-01-12",
      "title": "Fate/stay night: Heaven’s Feel II. lost butterfly",
      "video": false,
      "vote_average": 4.8,
      "vote_count": 13,
    },
    Object {
      "adult": false,
      "backdrop_path": "/juc9wt7Eh2IarLL5S1yQ1a21O1A.jpg",
      "genre_ids": Array [
        16,
        10751,
        35,
        14,
      ],
      "id": 360920,
      "original_language": "en",
      "original_title": "The Grinch",
      "overview": "The Grinch hatches a scheme to ruin Christmas when the residents of Whoville plan their annual holiday celebration.",
      "popularity": 135.309,
      "poster_path": "/gpkHvkCtZOeCQ2DelnJ2LB1WjZ5.jpg",
      "release_date": "2018-11-08",
      "title": "The Grinch",
      "video": false,
      "vote_average": 6.3,
      "vote_count": 671,
    },
    Object {
      "adult": false,
      "backdrop_path": "/ngBFDOsx13sFXiMweDoL54XYknR.jpg",
      "genre_ids": Array [
        53,
        27,
        9648,
        18,
      ],
      "id": 450465,
      "original_language": "en",
      "original_title": "Glass",
      "overview": "In a series of escalating encounters, security guard David Dunn uses his supernatural abilities to track Kevin Wendell Crumb, a disturbed man who has twenty-four personalities. Meanwhile, the shadowy presence of Elijah Price emerges as an orchestrator who holds secrets critical to both men.",
      "popularity": 133.566,
      "poster_path": "/svIDTNUoajS8dLEo7EosxvyAsgJ.jpg",
      "release_date": "2019-01-16",
      "title": "Glass",
      "video": false,
      "vote_average": 6.8,
      "vote_count": 851,
    },
    Object {
      "adult": false,
      "backdrop_path": "/7d6EY00g1c39SGZOoCJ5Py9nNth.jpg",
      "genre_ids": Array [
        28,
        12,
        16,
        878,
        35,
      ],
      "id": 324857,
      "original_language": "en",
      "original_title": "Spider-Man: Into the Spider-Verse",
      "overview": "Miles Morales is juggling his life between being a high school student and being a spider-man. When Wilson \\"Kingpin\\" Fisk uses a super collider, others from across the Spider-Verse are transported to this dimension.",
      "popularity": 128.098,
      "poster_path": "/laMM4lpQSh5z6KIBPwWogkjzBVQ.jpg",
      "release_date": "2018-12-07",
      "title": "Spider-Man: Into the Spider-Verse",
      "video": false,
      "vote_average": 8.5,
      "vote_count": 1370,
    },
    Object {
      "adult": false,
      "backdrop_path": "/r0ljJIPBGsGdpeDdlHpzCiNnxiB.jpg",
      "genre_ids": Array [
        18,
        53,
        9648,
        80,
        28,
      ],
      "id": 446807,
      "original_language": "en",
      "original_title": "The Girl in the Spider's Web",
      "overview": "In Stockholm, Sweden, hacker Lisbeth Salander is hired by Frans Balder, a computer engineer, to retrieve a program that he believes it is too dangerous to exist.",
      "popularity": 110.808,
      "poster_path": "/w4ibr8R702DCjwYniry1D1XwQXj.jpg",
      "release_date": "2018-10-26",
      "title": "The Girl in the Spider's Web",
      "video": false,
      "vote_average": 5.8,
      "vote_count": 246,
    },
    Object {
      "adult": false,
      "backdrop_path": "/wDN3FIcQQ1HI7mz1OOKYHSQtaiE.jpg",
      "genre_ids": Array [
        10751,
        14,
        12,
      ],
      "id": 338952,
      "original_language": "en",
      "original_title": "Fantastic Beasts: The Crimes of Grindelwald",
      "overview": "Gellert Grindelwald has escaped imprisonment and has begun gathering followers to his cause—elevating wizards above all non-magical beings. The only one capable of putting a stop to him is the wizard he once called his closest friend, Albus Dumbledore. However, Dumbledore will need to seek help from the wizard who had thwarted Grindelwald once before, his former student Newt Scamander, who agrees to help, unaware of the dangers that lie ahead. Lines are drawn as love and loyalty are tested, even among the truest friends and family, in an increasingly divided wizarding world.",
      "popularity": 107.868,
      "poster_path": "/kQKcbJ9uYkTQql2R8L4jTUz7l90.jpg",
      "release_date": "2018-11-14",
      "title": "Fantastic Beasts: The Crimes of Grindelwald",
      "video": false,
      "vote_average": 6.9,
      "vote_count": 3097,
    },
    Object {
      "adult": false,
      "backdrop_path": "/bOGkgRGdhrBYJSLpXaxhXVstddV.jpg",
      "genre_ids": Array [
        12,
        28,
        14,
      ],
      "id": 299536,
      "original_language": "en",
      "original_title": "Avengers: Infinity War",
      "overview": "As the Avengers and their allies have continued to protect the world from threats too large for any one hero to handle, a new danger has emerged from the cosmic shadows: Thanos. A despot of intergalactic infamy, his goal is to collect all six Infinity Stones, artifacts of unimaginable power, and use them to inflict his twisted will on all of reality. Everything the Avengers have fought for has led up to this moment - the fate of Earth and existence itself has never been more uncertain.",
      "popularity": 100.292,
      "poster_path": "/7WsyChQLEftFiDOVTGkv3hFpyyt.jpg",
      "release_date": "2018-04-25",
      "title": "Avengers: Infinity War",
      "video": false,
      "vote_average": 8.3,
      "vote_count": 11061,
    },
    Object {
      "adult": false,
      "backdrop_path": "/cwiJQXezWz876K3jS57Sq56RYCZ.jpg",
      "genre_ids": Array [
        14,
        10751,
        35,
      ],
      "id": 400650,
      "original_language": "en",
      "original_title": "Mary Poppins Returns",
      "overview": "In Depression-era London, a now-grown Jane and Michael Banks, along with Michael's three children, are visited by the enigmatic Mary Poppins following a personal loss. Through her unique magical skills, and with the aid of her friend Jack, she helps the family rediscover the joy and wonder missing in their lives.",
      "popularity": 99.319,
      "poster_path": "/uTVGku4LibMGyKgQvjBtv3OYfAX.jpg",
      "release_date": "2018-12-19",
      "title": "Mary Poppins Returns",
      "video": false,
      "vote_average": 6.8,
      "vote_count": 724,
    },
  ],
}
`;

exports[`returns the same state on an unhandled action 1`] = `
Object {
  "error": null,
  "isFetching": false,
  "page": 1,
  "results": Array [],
}
`;
